api/v1beta1/gcpmachine_types.go

```diff
@@ -146,6 +149,71 @@ const (
+	// "kmsKeyName": "projects/kms_project_id/locations/region/keyRings/key_region/cryptoKeys/key
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern=`projects\/[-_[A-Za-z0-9]+\/locations\/[-_[A-Za-z0-9]+\/keyRings\/[-_[A-Za-z0-9]+\/cryptoKeys\/[-_[A-Za-z0-9]+`
+	// +kubebuilder:validation:MaxLength=160
+	KMSKeyName string `json:"kmsKeyName,omitempty"`
+}
+
+// SuppliedKey contains a key for disk encryption. Either RawKey or RSAEncryptedKey must be provided.
+// +kubebuilder:validation:MinProperties=1
+type SuppliedKey struct {
+	// RawKey specifies a 256-bit customer-supplied encryption key, encoded in RFC 4648
+	// base64 to either encrypt or decrypt this resource. You can provide either the rawKey or the rsaEncryptedKey.
+	// For example: "rawKey": "SGVsbG8gZnJvbSBHb29nbGUgQ2xvdWQgUGxhdGZvcm0="
+	// +optional
+	RawKey []byte `json:"rawKey,omitempty"`
+	// RSAEncryptedKey specifies an RFC 4648 base64 encoded, RSA-wrapped 2048-bit customer-supplied encryption
+	// key to either encrypt or decrypt this resource. You can provide either the rawKey or the
+	// rsaEncryptedKey.
+	// For example: "rsaEncryptedKey": "ieCx/NcW06PcT7Ep1X6LUTc/hLvUDYyzSZPPVCVPTVEohpeHASqC8uw5TzyO9U+Fka9JFHi
+	// z0mBibXUInrC/jEk014kCK/NPjYgEMOyssZ4ZINPKxlUh2zn1bV+MCaTICrdmuSBTWlUUiFoDi
+	// D6PYznLwh8ZNdaheCeZ8ewEXgFQ8V+sDroLaN3Xs3MDTXQEMMoNUXMCZEIpg9Vtp9x2oe=="
+	// The key must meet the following requirements before you can provide it to Compute Engine:
+	// 1. The key is wrapped using a RSA public key certificate provided by Google.
+	// 2. After being wrapped, the key must be encoded in RFC 4648 base64 encoding.
+	// Gets the RSA public key certificate provided by Google at: https://cloud-certs.storage.googleapis.com/google-cloud-csek-ingress.pem
+	// +optional
+	RSAEncryptedKey []byte `json:"rsaEncryptedKey,omitempty"`
+}
+
+// CustomerEncryptionKey supports both Customer-Managed or Customer-Supplied encryption keys .
+type CustomerEncryptionKey struct {
+	// KeyType is the type of encryption key. Must be either Managed, aka Customer-Managed Encryption Key (CMEK) or
+	// Supplied, aka Customer-Supplied EncryptionKey (CSEK).
+	// +kubebuilder:validation:Enum=Managed;Supplied
+	KeyType KeyType `json:"keyType"`
+	// KMSKeyServiceAccount is the service account being used for the encryption request for the given KMS key.
+	// If absent, the Compute Engine default service account is used. For example:
+	// "kmsKeyServiceAccount": "name@project_id.iam.gserviceaccount.com.
+	// The maximum length is based on the Service Account ID (max 30), Project (max 30), and a valid gcloud email
+	// suffix ("iam.gserviceaccount.com").
+	// +kubebuilder:validation:MaxLength=85
+	// +kubebuilder:validation:XValidation:rule=`self.endsWith('.iam.gserviceaccount.com')`,message=`kmsKeyServiceAccount must be a valid gcloud service account. Must end with '.iam.gserviceaccount.com'.`
+	// +optional
+	KMSKeyServiceAccount *string `json:"kmsKeyServiceAccount,omitempty"`
```

Comment:
Non validation on the name or project id currently, could be done with CEL or with webhook for now

Reply:
Changed to use regexp instead of XValidation as multiple XValidations caused problems. This keeps all of the validations for this string in one place.

Reply:
Pattern works, curious what problems the multiple xvalidations added though if you have the output to hand

Reply:
Using the 3 XValidations had caused this error https://github.com/kubernetes-sigs/cluster-api-provider-gcp/pull/1137#discussion_r1491655487

Reply:
I suspect it would work if you include the 3 validations plus the maxlength on the field, from my read of that thread you haven't tried that combination yet

Reply:
Even with the `MaxLength` I get an error when using multiple XValidation rules, for example with just 2:
        // +kubebuilder:validation:MaxLength=85
       // +kubebuilder:validation:XValidation:rule="self.contains('@')",message="kmsKeyServiceAccount should be an email address"
       // +kubebuilder:validation:XValidation:rule="self.split('@')[0].matches('[a-z][-a-zA-Z0-9]{5,29}')",message="kmsServiceAccount email prefix must start with a lower case letter, followed by one or more lower case alphanumerical characters that can be separated by hyphens and must be between 6 and 30 characters in length."

I get:
```
  [FAILED] Unexpected error:
      <*fmt.wrapError | 0xc00061e180>: 
      unable to install CRDs onto control plane: unable to create CRD instances: unable to create CRD "gcpmachinetemplates.infrastructure.cluster.x-k8s.io": CustomResourceDefinition.apiextensions.k8s.io "gcpmachinetemplates.infrastructure.cluster.x-k8s.io" is invalid: [spec.validation.openAPIV3Schema.properties[spec].properties[template].properties[spec].properties[additionalDisks].items.properties[encryptionKey].properties[kmsKeyServiceAccount].x-kubernetes-validations[0].rule: Forbidden: estimated rule cost exceeds budget by factor of 3.7x (try simplifying the rule, or adding maxItems, maxProperties, and maxLength where arrays, maps, and strings are declared), spec.validation.openAPIV3Schema.properties[spec].properties[template].properties[spec].properties[additionalDisks].items.properties[encryptionKey].properties[kmsKeyServiceAccount].x-kubernetes-validations[1].rule: Forbidden: estimated rule cost exceeds budget by factor of 29.4x (try simplifying the rule, or adding maxItems, maxProperties, and maxLength where arrays, maps, and strings are declared), spec.validation.openAPIV3Schema.properties[spec].properties[template].properties[spec].properties[additionalDisks].items.properties[encryptionKey].properties[kmsKeyServiceAccount].x-kubernetes-validations[1].rule: Forbidden: contributed to estimated rule cost total exceeding cost limit for entire OpenAPIv3 schema, spec.validation.openAPIV3Schema.properties[spec].properties[template].properties[spec].properties[additionalDisks].items.properties[encryptionKey].properties[kmsKeyServiceAccount].x-kubernetes-validations[0].rule: Forbidden: contributed to estimated rule cost total exceeding cost limit for entire OpenAPIv3 schema, spec.validation.openAPIV3Schema: Forbidden: x-kubernetes-validations estimated rule cost total for entire OpenAPIv3 schema exceeds budget by factor of 3.3x (try simplifying the rule, or adding maxItems, maxProperties, and maxLength where arrays, maps, and strings are declared)]

```

Reply:
Cool, thanks for trying, lets continue with the validations we have here for now then

Reply:
Thanks